modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/splitter/QueryPlan.java

```diff
@@ -75,7 +85,21 @@ public void init(PlannerContext ctx) {
 
                 fragments.add(fragment);
 
-                parent.replaceInput(edge.childIndex(), new IgniteReceiver(cluster, traitSet, child.getRowType(), fragment));
+                if (parent != null)
+                    parent.replaceInput(edge.childIndex(), new IgniteReceiver(cluster, traitSet, child.getRowType(), fragment));
```

Comment:
Will this exception be thrown every time when the query is started on the client node? If so, can we add extra exchange before exception happened? Moreover I am not sure this exception-driven approach for mapping generation is a best way to implement it? Can we rethink current approach to avoid using exceptions?

Reply:
The exception will thrown if a client is a client node and calculated distribution is broadcast one. We may just cover this case, others are true exceptions.

Reply:
Ok, let's cover this case because IMO it is quite common unlike other cases when exception might be thrown (as I understand it right, it is possible only when some exotic affinity functions and node filters are used)

Reply:
OK, I'll create a ticket.

Reply:
https://issues.apache.org/jira/browse/IGNITE-12585